### PR TITLE
test(sampling): add seeded correctness fuzz coverage

### PR DESCRIPTION
**Pull request by JulianZJN**
## 📌 Description

Stream 1a of #3605: a seeded fuzzer for the existing probability-sampling entry points, following the MoE / GEMM fuzzers (`FuzzLedger`, env-var case count and seed, single-seed repro on failure). Test-only; no kernel or API changes.

Covered: `sampling_from_probs`, `top_k_sampling_from_probs`, `top_p_sampling_from_probs`, `top_k_top_p_sampling_from_probs` (`joint` and `top_k_first`), with scalar and per-request thresholds, `indices` (including `len(indices) != probs.size(0)`), and `seed`/`offset`/`generator` replay. `min_p`, chain speculative sampling and the logits entry points are left for a follow-up.

Oracle: a float64 PyTorch reference written from the documented set semantics (inclusive top-p nucleus; top-k keeps everything ≥ the k-th largest; `joint` = intersection of both masks; `top_k_first` = top-k renorm, then top-p). It does not call FlashInfer or mirror the rejection loop. Because the kernels accumulate in float32, the support check is two-sided with a tolerance of `vocab * 2^-24`: a draw fails only if it falls outside the loosened support, and every class inside the tightened support whose expected count is ≥ 64 must be drawn at least once (miss probability < e^-64). Ties at a top-k boundary are checked at support level only; no particular tie-break is asserted.

Cases: each seed fixes API, vocab (1–257), 1–4 distributions drawn from sparse / one-hot / bimodal / long-tail / near-uniform / strictly-decreasing / plateau shapes, k ∈ {1, 2, vocab−1, vocab}, and top-p either from a fixed grid or placed just below / at / just above one of the row's cumulative masses. ~4096 draws per case.

Distribution check: 12 frozen configs (including the no-op thresholds `top_k = vocab`, `top_p = 1`), 32768 draws each, Pearson χ² with low-expectation bins merged. A config fails only if `p < 1e-5 / 12` and the effect is large (TVD > 0.02 or max standardized residual > 6), so the family-wise false-positive budget is 1e-5 and a lone small p-value cannot fail CI. Thresholds were fixed before looking at FlashInfer output. Offline, over 200 seeds per config: 0 / 2400 gate firings for `torch.multinomial` and 0 / 2400 for FlashInfer; an injected bias of TVD 0.01 is caught 40 / 40 times (0.005: 15 / 40). A CPU-only test keeps the gate honest in both directions.

Found with it: per-request `top_k` combined with `indices` uses the wrong row's `k` (#5339, fix in #5340); `min_p` has the same pattern. That combination is left out of generation until the fix lands so this PR stays green on `main`; per-request `top_p` / `top_k_top_p` with `indices` are generated.

Cost: 254 tests, 1.4 s warm on one RTX 4090 D (first run JIT-compiles the sampling module). Registered next to `test_sampling.py` in `scripts/task_jit_run_tests_part3.sh`.

```
FLASHINFER_SAMPLING_FUZZ_NUM_CASES   number of seeded cases (default 240)
FLASHINFER_SAMPLING_FUZZ_SEED        base seed (default 0)
FLASHINFER_SAMPLING_FUZZ_ONLY_SEED   comma-separated seeds to replay
```

## 🔍 Related Issues

Part of #3605 (Stream 1a / Sampling). Related: #5339.

### ✅ Pre-commit Checks

- [x] I have installed `pre-commit` by running `pip install pre-commit` (or used your preferred method).
- [x] I have installed the hooks with `pre-commit install`.
- [x] I have run the hooks manually with `pre-commit run --all-files` and fixed any reported issues.

## 🧪 Tests

- [x] Tests have been added or updated as needed.
- [x] All tests are passing (`unittest`, etc.).

On RTX 4090 D (SM89), CUDA 12.8, torch 2.9.1, JIT build of `012542c8`:

- `pytest tests/utils/test_sampling_fuzz.py`: 254 passed; 5 fresh-process repeats, 254 passed each
- 1000 seeds split across two GPUs (`FLASHINFER_SAMPLING_FUZZ_SEED` / `NUM_CASES`): 1000 passed
- `pytest tests/utils/test_sampling.py -k 111`: 529 passed, 120 skipped (unchanged)

I only have SM89 here, so nothing above says anything about Hopper/Blackwell-specific paths.

## Reviewer Notes

The part most worth a second pair of eyes is the reference semantics in `reference_top_p_support` / `reference_top_k_support` and the float32 tolerance. If the intended contract differs anywhere (e.g. top-p exactly at a cumulative-mass boundary), I'd rather fix the oracle than encode the current behaviour.

## Summary by CodeRabbit

* **Tests**
  * Expanded coverage for probability-based sampling across supported filtering modes.
  * Added reproducibility checks for seeded sampling and generator replay.
  * Added validation for output types, index ranges, supported classes, and zero-probability handling.
  * Added statistical distribution checks against reference results, with automatic skipping when CUDA is unavailable.

### CODE REVIEW

File: tests/utils/test_sampling_fuzz.py [hunk 1]
```diff
@@ -0,0 +1,768 @@
...
+    # top_k_top_p: loosen both masks; top_k_first loosens the top-p of the renormalized row.
+    if order == "joint":
+        keep = reference_top_k_support(
+            probs_row, top_k, shift
+        ) & reference_top_p_support(probs_row, top_p, shift)
+        return keep & positive
+    renorm_k = reference_top_k_filter(probs_row, top_k)
+    keep = reference_top_k_support(probs_row, top_k, shift) & reference_top_p_support(
+        renorm_k, top_p, shift
+    )
```
[the remark this answers is not in the document]
Comment (JulianZJN) [reply tests/utils/test_sampling_fuzz.py]: Right, the top-p allowance had leaked into the top-k pivot, which dropped the k-th class from the must-draw side. Fixed in 6ccd49ff: top-k is compared exactly on the float32 values. I kept two masks rather than a single tie-inclusive one, because the docs only promise "keep the top-k probabilities" and `test_sampling.py` tolerates either choice at a tied pivot: may-draw is tie-inclusive, must-draw only requires a tied class when there is no freedom in choosing. `test_dropping_the_unique_kth_class_is_detected` covers the case the old oracle missed. The ledger is now consulted before any CUDA work as well.

File: tests/utils/test_sampling_fuzz.py [hunk 2]
```diff
@@ -0,0 +1,1168 @@
...
+    # Consulted before any CUDA work, so a quarantined config could never reach a
+    # kernel launch. Nothing is quarantined today; the entry below is a tolerated
+    # wrong answer, applied after the checks run.
+    _LEDGER.xfail_if_quarantined(cfg)
+    finding = _LEDGER.find(cfg)
+    try:
+        _run_case(cfg)
+    except _CaseFailure as exc:
+        if finding is not None:
+            pytest.xfail(f"[sampling] {cfg.label}: {finding.reason}\n{exc}")
```
[the remark this answers is not in the document]
Comment (JulianZJN) [reply tests/utils/test_sampling_fuzz.py:641]: Fixed in 3d8408f6. _CaseFailure now carries a category, and the ledger can xfail only a row_threshold failure whose observed samples match #5339's swapped-threshold signature. Shape, dtype, range, support-cardinality, seed-replay, and generator-replay failures remain hard failures. Added CPU tests for both waived and non-waived paths; the SM89 normal cases pass and the exact #5339 probe xfails as expected.
